hilti/runtime/src/types/bytes.cc

```diff
@@ -47,34 +106,86 @@ std::tuple<bool, Bytes::const_iterator> Bytes::find(const Bytes& needle, const c
 
 std::string Bytes::decode(unicode::Charset cs, unicode::DecodeErrorStrategy errors) const {
+    if ( Base::empty() )
+        return "";
+
     switch ( cs.value() ) {
         case unicode::Charset::UTF8: {
             std::string t;
 
-            auto p = reinterpret_cast<const unsigned char*>(Base::data());
-            auto e = p + Base::size();
+            auto p = Base::begin();
+            auto e = Base::end();
 
             while ( p < e ) {
-                utf8proc_int32_t cp;
-                auto n = utf8proc_iterate(p, e - p, &cp);
-
-                if ( n < 0 ) {
+                try {
+                    auto cp = utf8::next(p, e);
+                    utf8::append(cp, t);
+                } catch ( const utf8::exception& ) {
                     switch ( errors.value() ) {
                         case unicode::DecodeErrorStrategy::IGNORE: break;
-                        case unicode::DecodeErrorStrategy::REPLACE: t += "\ufffd"; break;
+                        case unicode::DecodeErrorStrategy::REPLACE: {
+                            utf8::append(unicode::REPLACEMENT_CHARACTER, t);
+                            break;
+                        }
                         case unicode::DecodeErrorStrategy::STRICT:
                             throw RuntimeError("illegal UTF8 sequence in string");
                     }
 
-                    p += 1;
-                    continue;
+                    p = std::next(p);
+                }
+            }
+
+            return t;
+        }
+
+        case unicode::Charset::UTF16BE: [[fallthrough]];
+        case unicode::Charset::UTF16LE: {
+            if ( Base::size() % 2 != 0 ) {
+                switch ( errors.value() ) {
+                    case unicode::DecodeErrorStrategy::STRICT: throw RuntimeError("illegal UTF16 character in string");
+                    case unicode::DecodeErrorStrategy::IGNORE: {
+                        // Ignore the last byte.
+                        return Bytes(str().substr(0, Base::size() / 2 * 2)).decode(cs, errors);
+                    }
+                    case unicode::DecodeErrorStrategy::REPLACE: {
+                        // Convert everything but the last byte, and append replacement.
+                        auto dec = Bytes(str().substr(0, Base::size() / 2 * 2)).decode(cs, errors);
+                        utf8::append(unicode::REPLACEMENT_CHARACTER, dec);
+                        return dec;
+                    }
                 }
+            }
+
+            // We can assume an even number of bytes.
 
-                t += std::string(reinterpret_cast<const char*>(p), n);
-                p += n;
+            std::u16string t;
+
+            // utfcpp expects to iterate a `u16string` or `u16string_view`.
+            auto v16 = std::u16string_view{reinterpret_cast<const char16_t*>(Base::data()), Base::size() / 2};
+
+            const auto order = cs.value() == unicode::Charset::UTF16LE ? U16Iterator::LE : U16Iterator::BE;
```

Comment:
For readability:

`- const auto order = cs.value() == unicode::Charset::UTF16LE ? U16Iterator::LE : U16Iterator::BE;`
`+ const auto order = (cs.value() == unicode::Charset::UTF16LE ? U16Iterator::LE : U16Iterator::BE);`